delta/plugins/project-deletion/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/projectdeletion/ProjectDeleter.scala

```diff
@@ -0,0 +1,99 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.projectdeletion
+
+import cats.syntax.all._
+import ch.epfl.bluebrain.nexus.delta.kernel.syntax.instantSyntax
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.IOUtils
+import ch.epfl.bluebrain.nexus.delta.plugins.projectdeletion.model.ProjectDeletionConfig
+import ch.epfl.bluebrain.nexus.delta.sdk.ProjectResource
+import monix.bio.UIO
+import ProjectDeleter._
+
+import java.time.{Duration, Instant}
+
+object ProjectDeleter {
+
+  def projectDeletionPass(
+      allProjects: UIO[Seq[ProjectResource]],
+      deleteProject: ProjectResource => UIO[Unit],
+      config: ProjectDeletionConfig,
+      lastEventTime: (ProjectResource, Instant) => UIO[Instant]
+  ): UIO[Unit] = {
+
+    val deleter = new ProjectDeleter(deleteProject, config, lastEventTime)
+
+    for {
+      allProjects <- allProjects
+      now         <- IOUtils.instant
+      _           <- allProjects.traverse(deleter.processProject(_, now))
+    } yield {
+      ()
+    }
+  }
+
+  implicit class BooleanTaskOps(val left: Boolean) extends AnyVal {
+    def `<||>`(right: UIO[Boolean]): UIO[Boolean] = {
+      if (left) {
+        UIO.pure(true)
+      } else {
+        right
+      }
+    }
+
+    def `<&&>`(right: UIO[Boolean]): UIO[Boolean] = {
+      if (left) {
+        right
+      } else {
+        UIO.pure(false)
+      }
+    }
+  }
+}
+
+class ProjectDeleter(
+    deleteProject: ProjectResource => UIO[Unit],
+    config: ProjectDeletionConfig,
+    lastEventTime: (ProjectResource, Instant) => UIO[Instant]
+) {
+
+  def processProject(pr: ProjectResource, now: Instant) = {
```

Comment:
I don't think this is a big deal. Especially because we don't have an integration test, I'm not sure of the value in this. The tests are already clear and have good coverage

Reply:
It could help reduce the boilerplate in the tests